add localization of quotients of multivarate polynomial rings to docu

wdecker opened the pull request:
After this PR has been merged, another PR by @HechtiDerLachs  will follow.

wdecker committed:
add localization of quotients of multivarate polynomial rings to docu

wdecker committed:
doctest corrections

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -15,45 +15,55 @@ Pages = ["localizations.md"]
-    Recall that in OSCAR, Gröbner bases are implemented for multivariate polynomial rings over fields (exact fields supported by OSCAR) and for multivariate polynomial rings over the integers.
+Write ``\frac{r}{u}`` for the equivalence class of ``(r, u)`` and ``R[U^{-1}]`` for the set of all equivalence classes.
+Mimicking the standard arithmetic for fractions, ``R[U^{-1}]`` can be made into a ring. This ring is called the
+*localization of ``R`` at ``U``*. It comes equipped with  the natural ring homomorphism
```
Do italics work well with mathematical expressions? Maybe they do, but we should check what this looks like in the final html.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -15,45 +15,55 @@ Pages = ["localizations.md"]
+then the preimage $U = p^{-1}(S)$ is a multiplicatively closed subset of ``R`` such that ``S = p(U)``.
+Creating the localization of the ring ``A`` at ``S `` in OSCAR amounts to localize the ``R``-module
+``A`` at ``U``, and equip the result with the appropriate ring structure. Hence, the corresponding
+constructor takes as input ``A`` and ``U`` rather than ``A`` and ``S``.
```
One nitpick is that this is not really what we're doing. For `A = R/I` and a multiplicative set `U` in `R` we have `p^{-1}(p(U)) = U + I` which is usually strictly bigger than the original `U`. By localizing `A` at `U` as an `R`-algebra, we effectively avoid having to worry about multiplicative sets of the form `U + I`! This is good, because they are in general much more difficult to handle than the original `U`s.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -15,45 +15,55 @@ Pages = ["localizations.md"]
-	
+    Most functions described in this section rely on the computation of standard bases. Recall that OSCAR supports standard
+    bases for multivariate polynomial rings over fields (exact fields supported by OSCAR) and for multivariate
+	polynomial rings over the integers.	
```
But for general localizations, the standard bases play only a minor role in a very particular subcase. I would probably not put this so prominently here.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
````diff
@@ -76,48 +86,20 @@ It is also possible to build products of multiplicatively closed sets already gi
-true
+Containment in multiplicatively closed subsets can be checked via the `in` function:
 
+```@docs
+in(f::MPolyElem, U::AbsMPolyMultSet)
````
This is more general: `in(a::RingElem, U::AbsMultSet)`.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
````diff
@@ -76,48 +86,20 @@ It is also possible to build products of multiplicatively closed sets already gi
 
 ```@docs
-Localization(U::AbsMPolyMultSet)
+localization(R::MPolyRing, U::AbsMPolyMultSet)
````
Same here: `Localization(R::Ring, U::AbsMultSet)`.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -128,40 +110,78 @@ If `Rloc` is the localization of a multivariate polynomial ring `R`  at a multip
 - `inverted_set(Rloc)` to `U`.
 
-###### Examples
+If `RQ` is a quotient of a multivariate polynomial ring `R`, `p : R -> RQ` is the projection map, `U`  is a
```
```suggestion
If `RQ` is a quotient of a multivariate polynomial ring `R`, ``p : R \to RQ`` is the projection map, `U`  is a
```

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -128,40 +110,78 @@ If `Rloc` is the localization of a multivariate polynomial ring `R`  at a multip
+`U` of `R`, then elements of `Rloc` are created as (fractions of) images of elements of `R` under
+the localization map or by coercing (pairs of) elements of `R` into fractions. 
+
+If `RQ` is a quotient of a multivariate polynomial ring `R`, `p : R -> RQ` is the projection map, `U`  is a
```
```suggestion
If `RQ` is a quotient of a multivariate polynomial ring `R`, ``p : R \to RQ`` is the projection map, `U`  is a
```

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -128,40 +110,78 @@ If `Rloc` is the localization of a multivariate polynomial ring `R`  at a multip
+the localization map or by coercing (pairs of) elements of `R` into fractions. 
+
+If `RQ` is a quotient of a multivariate polynomial ring `R`, `p : R -> RQ` is the projection map, `U`  is a
+multiplicatively closed subset of `R`, and `RQL` is the localization of `RQ` at `p(U)`, then elements of
```
```suggestion
multiplicatively closed subset of `R`, and `RQL` is the localization of `RQ` at `U`, then elements of
```
I would really write `U` here. You can localize `RQ` at `U` without ambiguity. And that's what we're doing.

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -224,45 +310,132 @@ localization of Multivariate Polynomial Ring in x, y, z over Rational Field at t
+- `S` be another ring.
+Then, to give a ring homomorphism `PHI`  from `Rloc` to `S` (from`RQL` to `S`) is the same
+as to give a ring homomorphism `phi` from `R` to `S` which sends elements of `U` to units
+in `S`. That is, `PHI` is determined by the composition of `PHI` with the localization map
```
```suggestion
in `S`. That is, `PHI` is determined by the composition of `phi` with the localization map
```

HechtiDerLachs commented on docs/src/CommutativeAlgebra/localizations.md:
````diff
@@ -224,45 +310,132 @@ localization of Multivariate Polynomial Ring in x, y, z over Rational Field at t
+projection map). The constructors below take this into account.
 
 ```@docs
 hom(Rloc::MPolyLocalizedRing, S::Ring, F::Map)
````
```suggestion
hom(Rloc::AbsLocalizedRing, S::Ring, F::Map)
```

wdecker committed:
small supplements

HechtiDerLachs commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -2328,45 +2353,60 @@ end
+the generators of `R`, and proceed as above.
+
+!!! warning
+    Except if `typeof(U) <: MPolyPowersOfElement` evaluates to `true`, the extra condion on `phi` is not checked by the function.
```
```suggestion
    Except from the case where the type of `U` is `<: MPolyPowersOfElement`, the extra condition on `phi` is not checked by this constructor.
```

HechtiDerLachs commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -266,6 +266,49 @@ function quo(
+    localization(RQ::MPolyQuo, U::AbsMPolyMultSet)
+
+Given a quotient `RQ` of a multivariate polynomial ring `R` with projection map
+`p : R -> RQ`, say, and given a multiplicative closed subset `U` of `R`, return the 
```
```suggestion
`p : R -> RQ`, say, and given a multiplicatively closed subset `U` of `R`, return the 
```

HechtiDerLachs reviewed (commented):
Over all, this looks OK to me. I left some minor points for discussion.

wdecker committed:
doctest correction, first reaction to code review

wdecker commented:
@HechtiDerLachs thx for your comments. I have reacted to some of your suggestions in the latest commit. There remain a few things which I wish to discuss with you. Unfortunately, I may not have time for this until Tuesday.

wdecker committed:
final polishing

wdecker commented:
From my point of view this is now ready to be merged if the test will pass. There will be a subsequent PR by @HechtiDerLachs with more on this.

HechtiDerLachs commented:
It seems that the current version is not even compiling. @wdecker Can you confirm this locally?

wdecker committed:
correction

fingolfin commented:
There are a bunch of unaddressed comments. But @wdecker explicitly asked me to merge this so work based on it can continue. So I'll do that now, and hope the remaining comments (such as changing `-> to `\to`) can be addressed later.

wdecker commented:
@fingolfin thx!
I think I addressed all comments except: 
1. the arrow discussion (if in a doctext almost every name is like `R`, `U` , that is, written as we write names in the docstring above the doctext, why should we have one thing in `` ...  ``? This looks strange to me.). But frankly, I do not care;
2. comments asking for more general types in some definitions of functions. It would be interesting to see suggestions for adjusting the corresponding doc tests.
